src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs

```diff
@@ -0,0 +1,223 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.SQLite.Interop;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    /// <summary>
+    /// Implementation of an <see cref="IPersistentStorage"/> backed by SQLite.
+    /// </summary>
+    internal partial class SQLitePersistentStorage : AbstractPersistentStorage
+    {
+        /// <summary>
+        /// Inside the DB we have a table dedicated to storing strings that also provides a unique 
+        /// integral ID per string.  This allows us to store data keyed in a much more efficient
+        /// manner as we can use those IDs instead of duplicating strings all over the place.  For
+        /// example, there may be many pieces of data associated with a file.  We don't want to 
+        /// key off the file path in all these places as that would cause a large amount of bloat.
+        /// 
+        /// Because the string table can map from arbitrary strings to unique IDs, it can also be
+        /// used to create IDs for compound objects.  For example, given the IDs for the FilePath
+        /// and Name of a Project, we can get an ID that represents the project itself by just
+        /// creating a compound key of those two IDs.  This ID can then be used in other compound
+        /// situations.  For example, a Document's ID is creating by compounding its Project's 
+        /// ID, along with the IDs for the Document's FilePath and Name.
+        /// 
+        /// The format of the table is:
+        /// 
+        ///  StringInfo
+        ///  ---------------------------------------------------------------
+        ///  | Id (integer, primary key, auto increment) | Value (varchar) |
+        ///  ---------------------------------------------------------------
+        /// </summary>
+        private const string StringInfoTableName = "StringInfo";
+
+        /// <summary>
+        /// Inside the DB we have a table for data corresponding to the <see cref="Solution"/>.  The 
+        /// data is just a blob that is keyed by a string Id.  Data with this ID can be retrieved
+        /// or overwritten.
+        /// 
+        /// The format of the table is:
+        /// 
+        ///  SolutionData
+        ///  --------------------------------------------
+        ///  | Id (primary key, varchar) | Value (blob) |
+        ///  --------------------------------------------
+        /// </summary>
+        private const string SolutionDataTableName = "SolutionData";
+
+        /// <summary>
+        /// Inside the DB we have a table for data that we want associated with a <see cref="Project"/>.
+        /// The data is keyed off of an integral value produced by combining the ID of the Project and
+        /// the ID of the name of the data (see <see cref="SQLitePersistentStorage.ReadStreamAsync(Project, string, CancellationToken)"/>.
```

Comment:
This cref did not help at all - I went through 4 layers of indirection before I actually saw how project ids were calculated.